presto-main/src/main/java/io/prestosql/failuredetector/CoordinatorGossipFailureDetector.java

```diff
@@ -0,0 +1,378 @@
+                                            FailureDetectorConfig failureDetectorConfig,
+                                            NodeInfo nodeInfo,
+                                            InternalCommunicationConfig internalCommunicationConfig,
+                                            GossipProtocolConfig config,
+                                            JsonCodec<WhomToGossipInfo> whomToGossipInfoCodec)
+    {
+        this(selector, httpClient, failureDetectorConfig, nodeInfo,
+                internalCommunicationConfig, config.getCnWkUpdateGossipMonitorServiceInterval(),
+                config.getCnGossipCollateInterval(), config.getGossipGroupSize(), whomToGossipInfoCodec);
+    }
+
+    private CoordinatorGossipFailureDetector(@ServiceType("presto") ServiceSelector selector,
+                                             @ForFailureDetector HttpClient httpClient,
+                                             FailureDetectorConfig failureDetectorConfig,
+                                             NodeInfo nodeInfo,
+                                             InternalCommunicationConfig internalCommunicationConfig,
+                                             Duration monitoringServiceUpdateInterval,
+                                             Duration cnGossipCollateInterval,
+                                             int gossipGroupSize,
+                                             JsonCodec<WhomToGossipInfo> whomToGossipInfoCodec)
+    {
+        super(selector, httpClient, failureDetectorConfig, nodeInfo, internalCommunicationConfig);
+        this.monitoringServiceUpdateInterval = monitoringServiceUpdateInterval;
+        this.gossipValidityPeriod = cnGossipCollateInterval;
+        this.whomToGossipInfoCodec = whomToGossipInfoCodec;
+        this.gossipGroupSize = gossipGroupSize;
+    }
+
+    public JsonCodec<WhomToGossipInfo> getWhomToGossipInfoCodec()
+    {
+        return whomToGossipInfoCodec;
+    }
+
+    @Override
+    protected MonitoringTask createNewTask(ServiceDescriptor service, URI uri)
+    {
+        log.debug("creating new gossip monitoring task for uri " + uri);
+        GossipMonitoringTask task = new GossipMonitoringTask(service, uri, uniq++);
+        tasks.add(task);
+        return task;
+    }
+
+    public List<URI> getSortedOnlineServiceDescriptors(Set<ServiceDescriptor> online)
+    {
+        List<URI> uris = new ArrayList<>();
+        List<ServiceDescriptor> onlineServices = new ArrayList<>(online);
+        Collections.sort(onlineServices, new Comparator<ServiceDescriptor>() {
+            @Override
+            public int compare(ServiceDescriptor t2, ServiceDescriptor t1)
+            {
+                return t1.getNodeId().compareTo(t1.getNodeId());
+            }
+        });
+        onlineServices.stream().forEach(o -> uris.add(getHttpUri(o)));
+        return uris;
+    }
+
+    private int[] assignLocalGossipGroups(int idx, int clusterSize)
```

Comment:
Cljuster size may keep changing thereby causing Indices to become unreliable... better to maintain node-id to ordinal-number map for repeated reference sanity

Reply:
it is always a pair of ints, containing the start and end index. Nodes are taken as a sorted list. start and end index will give howmuch ever nodes are there for the particular gossip group.